### DISCUSSION TITLE
Add filter to simulation output

### DISCUSSION THREAD

**Discussion by FMaxSokol** [cocotb/cocotb#4145] [Q&A] [answered]
Hello,

i am struggling currently...
I am getting NUMERIC_STD warnings because of meta values from my Questa simulation. I know where they are coming from and would normally work on my HDL to get rid of them.
However in this case I am not able to get rid of them in HDL, so they are expected behavior.

This results in the Questa simulation spamming the output with these error messages: `** Warning: NUMERIC_STD."=": metavalue detected, returning FALSE`.
I am trying to add a filter to the logging module to suppress these messages (simply by suppressing every message that includes `NUMERIC_STD` for now, later it will be changed to suppress warnings from the specific module) using the following filter:

```
class NumericStdFilter(logging.Filter):
    def filter(self, record: logging.LogRecord) -> bool:
        # Return False to filter out the log message if it contains "NUMERIC_STD"
        return "NUMERIC_STD" not in record.getMessage()
```

This filter works flawlessly when applied to the `_log` of my DUT.
When trying to log any message containing `NUMERIC_STD`, it does not get forwarded to stdout.

However, this does not suppress the messages from the simulation output, and I am unable to get the logger that outputs the simulation messages.
I even tried to simply add the filter to every logger currently available:

```
# get all loggers currently available
loggers = [logging.getLogger(name) for name in logging.getLogger().manager.loggerDict]

for logger_i in loggers:
    logger_i.addFilter(NumericStdFilter())
```

I do not want to turn off the numeric std warnings in the modelsim.ini, as I only want to suppress these warnings from this specific module. :)
I am using the current stable branch of cocotb (1.8.1).

Is anyone able to tell me, where I can find the logger that outputs the simulation messages, or does anyone have a better approach than I currently have?

**Comment by ktbarrett** [comment, marked as answer]
That logger is built into the simulator. You'll have to consult the user manual of your simulator or by conversing with the simulator provider.

**Reply by FMaxSokol** [reply]
Okay, I see... I thought this must go through python at some point, where I might be able to filter it.

Thank you very much for your support though :)
